### PR TITLE
xftp: negotiate protocol with ALPN

### PR DESCRIPTION
**Pull request by dpwiz**

### CODE REVIEW

File: src/Simplex/FileTransfer/Client.hs [hunk 1]
```diff
@@ -85,26 +93,64 @@ defaultXFTPClientConfig :: XFTPClientConfig
...
-  let HTTP2Client {sessionId} = http2Client
-      thParams = THandleParams {sessionId, blockSize = xftpBlockSize, thVersion = currentXFTPVersion, thAuth = Nothing, implySessId = False, batch = True}
-      c = XFTPClient {http2Client, thParams, transportSession, config}
+  http2Client <- liftError' xftpClientError $ getVerifiedHTTP2Client (Just username) useHost usePort (Just keyHash) Nothing http2Config clientDisconnected
+  let HTTP2Client {sessionId, sessionALPN} = http2Client
+      thParams0 = THandleParams {sessionId, blockSize = xftpBlockSize, thVersion = VersionXFTP 1, thAuth = Nothing, implySessId = False, batch = True}
+  logDebug $ "Client negotiated handshake protocol: " <> tshow sessionALPN
+  thParams <- case fromMaybe "xftp/0" sessionALPN of
+    "xftp/0" -> pure thParams0
+    "xftp/1" -> xftpClientHandshakeV1 g serverVRange keyHash http2Client thParams0
```
Comment (epoberezkin) [inline src/Simplex/FileTransfer/Client.hs:116]:
```suggestion
  thParams <- case sessionALPN of
    Just "xftp/1" -> xftpClientHandshakeV1 g serverVRange keyHash http2Client thParams0
    Nothing -> pure thParams0
```

File: src/Simplex/FileTransfer/Client.hs [hunk 2]
```diff
@@ -85,26 +93,64 @@ defaultXFTPClientConfig :: XFTPClientConfig
...
+    _ -> throwError $ PCETransportError (TEHandshake VERSION)
+  let c = XFTPClient {http2Client, thParams, transportSession, config}
   atomically $ writeTVar clientVar $ Just c
   pure c
 
+xftpClientHandshakeV1 :: TVar ChaChaDRG -> VersionRangeXFTP -> C.KeyHash -> HTTP2Client -> THandleParamsXFTP -> ExceptT XFTPClientError IO THandleParamsXFTP
+xftpClientHandshakeV1 g serverVRange keyHash@(C.KeyHash kh) c@HTTP2Client {sessionId, serverKey, serverCerts = _todo} thParams0 = do
+  -- step 1: pass initiative to server to get its handshake
```
Comment (epoberezkin) [inline src/Simplex/FileTransfer/Client.hs:124]: that's a symptom that it would be helpful to have 4 functions with the names

File: src/Simplex/FileTransfer/Server.hs [hunk 3]
```diff
@@ -92,12 +102,58 @@ xftpServer cfg@XFTPServerConfig {xftpPort, transportConfig, inactiveClientExpira
...
+        runHTTP2Server started xftpPort defaultHTTP2BufferSize serverParams transportConfig inactiveClientExpiration cleanup $ \sessionId sessionALPN r sendResponse -> do
+          logDebug $ "Serving session " <> tshow (B.take 5 $ strEncode sessionId) <> " request using " <> tshow sessionALPN
           reqBody <- getHTTP2Body r xftpBlockSize
-          let thParams = THandleParams {sessionId, blockSize = xftpBlockSize, thVersion = currentXFTPVersion, thAuth = Nothing, implySessId = False, batch = True}
-          processRequest XFTPTransportRequest {thParams, request = r, reqBody, sendResponse} `runReaderT` env
+          let thParams0 = THandleParams {sessionId, blockSize = xftpBlockSize, thVersion = VersionXFTP 1, thAuth = Nothing, implySessId = False, batch = True}
+              req0 = XFTPTransportRequest {thParams = thParams0, request = r, reqBody, sendResponse}
+          flip runReaderT env $ case fromMaybe "xftp/0" sessionALPN of
+            "xftp/0" -> processRequest req0
+            "xftp/1" ->
```
Comment (epoberezkin) [inline src/Simplex/FileTransfer/Server.hs:118]:
```suggestion
          flip runReaderT env $ case sessionALPN of
            Nothing -> processRequest req0
            Just "xftp/1" ->
```

File: src/Simplex/FileTransfer/Protocol.hs [hunk 4]
```diff
@@ -65,7 +65,7 @@ xftpBlockSize :: Int
 xftpBlockSize = 16384
 
 -- | File protocol clients
-data FileParty = FRecipient | FSender
+data FileParty = FRecipient | FSender -- XXX: FHandshake?
```
Comment (epoberezkin) [inline src/Simplex/FileTransfer/Protocol.hs:68]: .

File: src/Simplex/FileTransfer/Protocol.hs [hunk 5]
```diff
@@ -244,6 +244,7 @@ data FileResponseTag
   | FROk_
   | FRErr_
   | FRPong_
+  | FRHandshake_
```
Comment (epoberezkin) [inline src/Simplex/FileTransfer/Protocol.hs:247]: ?

File: src/Simplex/FileTransfer/Client.hs [hunk 6]
```diff
@@ -85,26 +93,64 @@ defaultXFTPClientConfig :: XFTPClientConfig
...
+  let helloReq = H.requestNoBody "POST" "/" []
+  HTTP2Response {respBody = HTTP2Body {bodyHead = shsBody'}} <- liftError' (const $ PCEResponseError HANDSHAKE) $ sendRequestDirect c helloReq Nothing
+  shsBody <- liftEitherWith (const $ PCEResponseError HANDSHAKE) $ C.unPad shsBody'
+  XFTPServerHandshake {xftpVersionRange, sessionId = serverSessId, authPubKey = serverAuth} <- liftEitherWith (const $ PCEResponseError HANDSHAKE) $ smpDecode shsBody
```
Comment (epoberezkin) [inline src/Simplex/FileTransfer/Client.hs:128]: it doesn't allow extending headers, partial parsers should be allowed

File: src/Simplex/FileTransfer/Client.hs [hunk 7]
```diff
@@ -85,26 +93,64 @@ defaultXFTPClientConfig :: XFTPClientConfig
...
+      pure (thParams0 {thAuth = Just auth, thVersion = v}, XFTPClientHandshake {xftpVersion = v, keyHash, authPubKey = pk})
+  -- step 3: send client handshake
+  let chsReq = H.requestBuilder "POST" "/" [] $ padXftp (smpEncode chs)
+  HTTP2Response {respBody = HTTP2Body {bodyHead}} <- liftError' (const $ PCEResponseError HANDSHAKE) $ sendRequestDirect c chsReq Nothing
+  -- step 4: validate server confirmation
+  unless (B.null bodyHead) $ throwError $ PCEResponseError HANDSHAKE
+  -- all done, return negotiated params
+  pure params
+  where
+    padXftp bs = either (error "assert: xftpBlockSize < 65k") byteString $ C.pad bs xftpBlockSize
```
Comment (epoberezkin) [inline src/Simplex/FileTransfer/Client.hs:152]: why don't you want to keep it as failable? It's in ExceptT anyway

File: src/Simplex/FileTransfer/Client.hs [hunk 8]
```diff
@@ -198,10 +244,11 @@ downloadXFTPChunk g c@XFTPClient {config} rpKey fId chunkSpec@XFTPRcvChunkSpec {
         let t = chunkTimeout config chunkSize
         ExceptT (sequence <$> (t `timeout` download cbState)) >>= maybe (throwError PCEResponseTimeout) pure
         where
-          download cbState = runExceptT $
-            withExceptT PCEResponseError $
-              receiveEncFile chunkPart cbState chunkSpec `catchError` \e ->
-                whenM (doesFileExist filePath) (removeFile filePath) >> throwError e
+          download cbState =
+            runExceptT $
+              withExceptT PCEResponseError $
```
Comment (epoberezkin) [inline src/Simplex/FileTransfer/Client.hs:249]:
```suggestion
            runExceptT . withExceptT PCEResponseError $
```

File: src/Simplex/FileTransfer/Client.hs [hunk 9]
```diff
@@ -198,10 +244,11 @@ downloadXFTPChunk g c@XFTPClient {config} rpKey fId chunkSpec@XFTPRcvChunkSpec {
...
+                receiveEncFile chunkPart cbState chunkSpec `catchError` \e ->
+                  whenM (doesFileExist filePath) (removeFile filePath) >> throwError e
```
Comment (epoberezkin) [inline src/Simplex/FileTransfer/Client.hs:251]: why this change?

File: src/Simplex/FileTransfer/Server/Env.hs [hunk 10]
```diff
@@ -104,7 +106,16 @@ newXFTPServerEnv config@XFTPServerConfig {storeLogFile, fileSizeQuota, caCertifi
...
     logInfo $ "Total / available storage: " <> tshow quota <> " / " <> tshow (quota - used)
     when (quota < used) $ logInfo "WARNING: storage quota is less than used storage, no files can be uploaded!"
-  tlsServerParams <- liftIO $ loadTLSServerParams caCertificateFile certificateFile privateKeyFile
+  tlsServerParams' <- liftIO $ loadTLSServerParams caCertificateFile certificateFile privateKeyFile
+  let tlsServerParams =
+        tlsServerParams'
+          { T.serverHooks =
+              def
+                { T.onALPNClientSuggest = Just $ \case
+                    [v1@"xftp/1"] -> pure v1
```
Comment (epoberezkin) [inline src/Simplex/FileTransfer/Server/Env.hs:115]: what if client sends "xftp/2" too? shouldn't "xftp/1" be one of the strings?

File: src/Simplex/FileTransfer/Client.hs (hunk 7 above)
Comment (dpwiz) [reply src/Simplex/FileTransfer/Client.hs:152]: request building is pure, didn't want to interleave with never-firing handlers.

this place would light up when searching for `xftpBlockSize`-affected code

File: src/Simplex/Messaging/Transport/HTTP2/Client.hs [hunk 11]
```diff
@@ -109,10 +114,20 @@ getVerifiedHTTP2ClientWith config host port disconnected setup =
...
-    client c cVar sessionId sendReq = do
+    client :: HClient -> TMVar (Either HTTP2ClientError HTTP2Client) -> TLS -> H.Client HTTP2Response
+    client c cVar tls sendReq = do
       sessionTs <- getCurrentTime
-      let c' = HTTP2Client {action = Nothing, client_ = c, sendReq, sessionId, sessionTs}
+      let c' =
+            HTTP2Client
+              { action = Nothing,
+                client_ = c,
+                serverKey = either (error "assert: TLS has server chain and key") id $ getServerVerifyKey tls,
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Transport/HTTP2/Client.hs:124]: will this leak on error?

File: src/Simplex/FileTransfer/Client.hs [hunk 12]
```diff
@@ -85,26 +93,63 @@ defaultXFTPClientConfig :: XFTPClientConfig
...
+xftpClientHandshakeV1 g serverVRange keyHash@(C.KeyHash kh) c@HTTP2Client {sessionId, serverKey} thParams0 = do
+  shs <- getServerHandshake
+  (v, sk) <- processServerHandshake shs
+  (pk, k) <- atomically $ C.generateKeyPair g
+  sendClientHandshake XFTPClientHandshake {xftpVersion = v, keyHash, authPubKey = pk}
+  pure thParams0 {thAuth = Just THandleAuth {peerPubKey = sk, privKey = k}, thVersion = v}
+  where
+    getServerHandshake = do
+      let helloReq = H.requestNoBody "POST" "/" []
+      HTTP2Response {respBody = HTTP2Body {bodyHead = shsBody'}} <- liftError' (const $ PCEResponseError HANDSHAKE) $ sendRequestDirect c helloReq Nothing
```
Comment (epoberezkin) [inline src/Simplex/FileTransfer/Client.hs:132]: this assumes that the first chunk of the response is the full response and the rest is not received
